Marlin/src/module/tool_change.cpp

```diff
@@ -494,225 +476,219 @@ void tool_change(const uint8_t tmp_extruder, const float fr_mm_s/*=0.0*/, bool n
-              current_position[Z_AXIS] += MAX(-zdiff, 0.0) + toolchange_settings.z_raise;
-              planner.buffer_line(current_position, planner.settings.max_feedrate_mm_s[Z_AXIS], active_extruder);
-              move_nozzle_servo(tmp_extruder);
-            #endif
+            return;
           #endif
+        }
+        else {
+      #endif
 
-          #if ENABLED(DEBUG_LEVELING_FEATURE)
-            if (DEBUGGING(LEVELING)) {
-              SERIAL_ECHOPAIR("Offset Tool XY by { ", xdiff);
-              SERIAL_ECHOPAIR(", ", ydiff);
-              SERIAL_ECHOPAIR(", ", zdiff);
-              SERIAL_ECHOLNPGM(" }");
-            }
-          #endif
+      if (toolchange_settings.swap_length) {
+        #if ENABLED(ADVANCED_PAUSE_FEATURE)
+          do_pause_e_move(-toolchange_settings.swap_length, MMM_TO_MMS(toolchange_settings.retract_speed));
+        #else
+          current_position[E_AXIS] -= toolchange_settings.swap_length / planner.e_factor[active_extruder];
+          planner.buffer_line(current_position, MMM_TO_MMS(toolchange_settings.retract_speed), active_extruder);
+        #endif
+      }
+      #if ENABLED(PREVENT_COLD_EXTRUSION)
+        }
+      #endif
+    #endif // TOOLCHANGE_FILAMENT_SWAP
 
-          // The newly-selected extruder XY is actually at...
-          current_position[X_AXIS] += xdiff;
-          current_position[Y_AXIS] += ydiff;
-          current_position[Z_AXIS] += zdiff;
+    if (tmp_extruder != active_extruder) {
+      const float old_feedrate_mm_s = fr_mm_s > 0.0 ? fr_mm_s : feedrate_mm_s;
+      feedrate_mm_s = fr_mm_s > 0.0 ? fr_mm_s : XY_PROBE_FEEDRATE_MM_S;
 
-          // Set the new active extruder
+      #if ENABLED(DUAL_X_CARRIAGE)
+        #if HAS_SOFTWARE_ENDSTOPS && HAS_WORKSPACE_OFFSET
+          // Update the X software endstops early
           active_extruder = tmp_extruder;
+          update_software_endstops(X_AXIS);
+          active_extruder = !tmp_extruder;
 
-        #endif // !DUAL_X_CARRIAGE
-
-        // Tell the planner the new "current position"
-        sync_plan_position();
-
-        #if ENABLED(DELTA)
-          //LOOP_XYZ(i) update_software_endstops(i); // or modify the constrain function
-          const bool safe_to_move = current_position[Z_AXIS] < delta_clip_start_height - 1;
+          // Don't move the new extruder out of bounds
+          if (!WITHIN(current_position[X_AXIS], soft_endstop_min[X_AXIS], soft_endstop_max[X_AXIS]))
+            no_move = true;
         #else
-          constexpr bool safe_to_move = true;
+            // No software endstops? Use the configured limits
+            if (active_extruder == 0) {
+              if (!WITHIN(current_position[X_AXIS], X2_MIN_POS, X2_MAX_POS))
+                no_move = true;
+            }
+            else if (!WITHIN(current_position[X_AXIS], X1_MIN_POS, X1_MAX_POS))
+              no_move = true;
         #endif
+      #endif
+      if( !no_move ) {
+        if (!no_move) set_destination_from_current();
```

Comment:
No need to check this inside the `if (!no_move)` block.

Reply:
Note that `do_blocking_move_to...` functions can and will overwrite `destination`.